src/cluster/session.rs

```diff
@@ -150,6 +149,18 @@ impl<
 {
 }
 
+#[async_trait]
+impl <LB> ResponseCache for Session<LB> where LB: Send {
+    async fn match_or_cache_response(&self, stream_id: i16, frame: Frame) -> Option<Frame> {
+        if frame.stream == stream_id {
```

Comment:
Here we may want to check if `self.responses` already have a response with an expected stream id. It can happen that it already among of responses because it was received during one of previous attempts.

Reply:
This is already done in line 160, although I need to fix that double lock.

Reply:
Indeed 👍